### ISSUE TITLE
Aim of the B_Damage module

### ISSUE THREAD

**Issue by JonnyCBB** [td93/B_Damage#2]
So I thought I would raise this as an issue because I think we should change the goal of this B_Damage repo.   

When we wrote the MATLAB code it was written in a procedural manner and the end result was a text file that essentially was the "ATOM" lines of a PDB file with additional columns for B_Damage related info. But I'm not so sure that the B_Damage package should do this in one script because I think we can make it more useful.    

A piece of evidence that suggests we can do better are the scripts that exist in the "AnalyseProgressiveBdamage" directory. The fact that if you want to do anything with the B_Damage values you have to parse the text file again and then do some analysis.

A better use of this package would be to make it a library that can calculate B_Damage values, but it also allows you to do more analysis via methods that are defined and usable from the library.    

For example. Let's suppose I want to calculate B_Damage values for a deposited structure with PDB code "2BN3", then we would write something like:

``` python
import B_Damage as bdam
pdb = bdam.parse_pdb('2bn3')
pdb.calculate_b_damage()
```

Notice that we are able to parse the pdb file without having to calculate B_Damage. This is in case the user may want to analyse the pdb info without necessarily going through the lengthy steps for calculating B_Damage. E.g. they might want to do some analysis with the standard atomic B factors first.   

At this point we have created an object with variable name `pdb` (we should decide what this object should contain and what it should do) and this object contains atoms and their associated B_Damage values.  
Now if I wanted to I could plot the B_Damage values as follows:

``` python
pdb.plot_bdamage()
```

Or I could write out the B_Damage values to a file with

``` python
pdb.write_bdamage_to_file()
```

These are just a few of the things that would be possible. You may also want to calculate B_Damage Ratio amongst other things.   

Basically the idea is that B_Damage code isn't just a single procedural script but rather it's a library that is flexible and allows you to do all of the analysis much faster and easier and it's all written in.   

What do you think?

**Comment by JonnyCBB** [comment]
@kls93 I think that we can make the output of B Damage slightly more flexible and in line with what I've mentioned above.    

At the moment the program can be run from the command line with:

```
>python cambda.py input_file.txt
```

and this spits out some files which people can look at and use. However if I want to plot a histogram with a different `threshold` I would have to run the **entire** program again and change the input file. This can be pretty tedious to do.   

I think we should aim to have write something so that `cambda` is more than just a method but is a class that people can use. For example if I open a Python console we could do something like:

``` python
my_cambda_object = cambda("2BN3", PDT=14, threshold=0.05):

# I can access the dataframe like this and do all the usual dataframe stuff
my_cambda_object.df
```

Notice that now I can do various things with the dataframe. When the `cambda` method is run, it can produce all of the files that are currently produced so it still is easy for general users to get what they want. However, for more interested and tech savvy users they should also be allowed to make a histogram again by doing something like:

``` python
my_cambda_object.make_histogram(fileName="my_hist.png")
```

This allows someon to plot a new histogram without having to rerun the entire B damage function (which can take a while for large structures as you already know).   

Basically for developers/people studying radiation damage I think this would be much more convenient for them to explore exactly what it is they want whilst also allowing general users to get the information they need very simply. I hope this makes sense.

**Comment by kls93** [comment]
@JonnyCBB 
Hi Jonny,

Currently I have split the programme into three:
- writing python rabdam_dataframe.py INPUT.txt into the command line returns a dataframe which is saved as a pkl file in the logfiles folder. A developer could then open the dataframe with pd.read_pickle in a Python console and do whatever they want with the data
- writing python rabdam_py INPUT.txt into the command line runs the complete programme, returning the current output files (first rabdam_dataframe is run, then rabdam_analysis)
- writing python rabdam_analysis.py INPUT.txt into the command line will write the current output files from a dataframe input (identified by the PDB code), so you can change the analysis parameters without having to calculate the Bdamage values again

(Note that the same INPUT file is used to run all three programmes.)

What do you think? Is this sufficiently flexible for both a user and a developer?

**Comment by JonnyCBB** [comment]
Hi @kls93 It's not exactly what I had in mind but I think this will do. It's better than the previous solution we had :+1:

**Comment by JonnyCBB** [comment]
There have been enough changes to the code since I posted this that I think it's best that I close this issue. If we do go down the web server route then this should abstract away much of the issues raised here.
